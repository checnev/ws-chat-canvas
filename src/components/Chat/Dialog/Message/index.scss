.message {
  display: flex;
  flex-direction: column;
  margin-bottom: 15px;

  &__text {
    display: inline-block;
    padding: 8px;
    margin-bottom: 10px;
    border: 1px solid rgba(0, 0, 0, .1);
    border-radius: 4px;
    background:rgb(255, 255, 255);
    white-space: pre-line;
  }

  &__date {
    color: rgb(153, 153, 153);
    font-size: .8em;
  }

  &_direction {
    &_in {
      align-items: flex-start;
    }

    &_out {
      align-items: flex-end;

      .userInfo {
        display: none;
        justify-content: flex-end;

        &__avatar {
          order: 1;
        }

        &__username {
          margin-left: 0;
          margin-right: 10px;
        }
      }
    }
  }
}
