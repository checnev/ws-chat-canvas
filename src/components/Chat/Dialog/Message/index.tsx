import React from 'react';
import cn from 'utils/cn';
import UserInfo from 'components/UserInfo';
import type { IUser } from 'lib/user/User';

import './index.scss';

export interface IMessage {
  id: number;
  author: IUser;
  text: string;
  date: Date;
}

interface MessageProps {
  message: IMessage;
  direction: 'in' | 'out';
}

class Message extends React.Component<MessageProps> {
  render() {
    const message = cn('message');
    const { author, date, text } = this.props.message;
    return (
      <div
        className={
          message({ direction: this.props.direction })
        }
      >
        <div className={message('text')}>
          {text}
        </div>
        <div className={message('author')}>
          <UserInfo user={author} />
          <span className={message('date')}>
            {date.toLocaleString().replace(',', '')}
          </span>
        </div>
      </div>
    );
  }
}

export default Message;
