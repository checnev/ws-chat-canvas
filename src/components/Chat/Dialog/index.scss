.dialog {
  position: relative;
  flex-grow: 1;
  min-height: 100px;

  &__window {
    position: absolute;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    width:100%;
    height:100%;
    top: 0;
    left: 0;
  }

  &__messages {
    overflow-y: auto;
    padding: 10px;
  }
}
