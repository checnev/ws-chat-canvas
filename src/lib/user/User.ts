import { Color } from 'constants/colors';

export interface IUser {
  name: string;
  color: Color;
  id?: string;
}

class User implements IUser {
  constructor(
    public readonly id: string,
    public readonly name: string,
    public color: Color
  ) {}
  
}

export default User;
