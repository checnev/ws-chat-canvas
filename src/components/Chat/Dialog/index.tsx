import React from 'react';
import cn from 'utils/cn';
import Message, { IMessage } from './Message';
import './index.scss';
import type User from 'lib/user/User';

interface DialogProps {
  messages: IMessage[];
  user: User;
}

class Dialog extends React.Component<DialogProps> {
  messagesRef = React.createRef<HTMLDivElement>();

  componentDidUpdate() {
    const messages = this.messagesRef.current;

    messages?.scrollBy({
      top: messages.scrollHeight,
      behavior: 'smooth',
    });
  }

  render() {
    const dialog = cn('dialog');

    return (
      <div className={dialog()}>
        <div className={dialog('window')}>
          <div className={dialog('messages')} ref={this.messagesRef}>
            {this.props.messages.map((message) => (
              <Message
                key={message.id}
                direction={message.author.id === this.props.user.id ? 'out' : 'in'}
                message={message}
              />
            ))}
          </div>
        </div>
      </div>
    );
  }
}

export default Dialog;
