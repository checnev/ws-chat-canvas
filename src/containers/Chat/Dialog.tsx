import React from 'react';
import { UserContext } from 'context/user';
import Dialog from 'components/Chat/Dialog';
import type { IMessage } from 'components/Chat/Dialog/Message';
import type Request from 'lib/api/Request';

interface DialogContainerState {
  messages: IMessage[];
}

class DialogContainer extends React.Component<{}, DialogContainerState> {
  static contextType = UserContext;
  context!: React.ContextType<typeof UserContext>

  state = {
    messages: [],
  };

  recieveMessage = (req: Request) => {
    this.setState((state) => {
      const message = req.data;
      message.date = new Date(message.date);

      return { messages: [...state.messages, req.data] };
    });
  };

  componentDidMount() {
    this.context.api?.subscribe('chat', this.recieveMessage);
  }

  componentWillUnmount() {
    this.context.api?.unsubscribe('chat', this.recieveMessage);
  }

  render() {
    return <Dialog messages={this.state.messages} user={this.context.user!} />;
  }
}

export default DialogContainer;
