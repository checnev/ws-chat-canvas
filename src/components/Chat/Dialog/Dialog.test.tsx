import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import User from 'lib/user/User';
import Dialog from 'components/Chat/Dialog';

let container: any = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const naruto = new User('1', 'Naruto', 'green');
const sasuke = new User('2', 'Sasuke', 'orange');

const messages = [
  {
    id: 1,
    author: naruto,
    text: 'Come back to the village!',
    date: new Date(2021, 0, 30, 14),
  },
  {
    id: 2,
    author: sasuke,
    text: 'No!',
    date: new Date(2021, 0, 30, 15),
  }
];

it('should render Dialog', () => {
  act(() => {
    render(<Dialog messages={messages} user={naruto} />, container);
  });

  expect(
    pretty(container.innerHTML)
  ).toMatchSnapshot();
});
