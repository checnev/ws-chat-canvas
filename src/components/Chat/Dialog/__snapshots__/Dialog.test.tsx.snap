// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Dialog 1`] = `
"<div class=\\"dialog\\">
  <div class=\\"dialog__window\\">
    <div class=\\"dialog__messages\\">
      <div class=\\"message message_direction_out\\">
        <div class=\\"message__text\\">Come back to the village!</div>
        <div class=\\"message__author\\">
          <div class=\\"userInfo\\">
            <div class=\\"userInfo__avatar\\" style=\\"background-color: green;\\">N</div><span class=\\"userInfo__username\\">Naruto</span>
          </div><span class=\\"message__date\\">30.01.2021 14:00:00</span>
        </div>
      </div>
      <div class=\\"message message_direction_in\\">
        <div class=\\"message__text\\">No!</div>
        <div class=\\"message__author\\">
          <div class=\\"userInfo\\">
            <div class=\\"userInfo__avatar\\" style=\\"background-color: orange;\\">S</div><span class=\\"userInfo__username\\">Sasuke</span>
          </div><span class=\\"message__date\\">30.01.2021 15:00:00</span>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
